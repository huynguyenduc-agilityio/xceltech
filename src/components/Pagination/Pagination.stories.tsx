import type { Meta, StoryObj } from '@storybook/react';

// Components
import Pagination from '.';

const meta: Meta<typeof Pagination> = {
  title: 'Components/Pagination',
  component: Pagination,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
  render: () => (
    <div className="flex justify-center w-[1000px] p-9 bg-blue-light">
      <Pagination />
    </div>
  ),
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  args: {
    totalRecords: 200,
    limit: 10,
    currentPage: 2,
  },
};
