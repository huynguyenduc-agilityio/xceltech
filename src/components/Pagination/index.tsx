import { memo } from 'react';
import isEqual from 'react-fast-compare';

// Icons
import { MoveLeft, MoveRight } from 'lucide-react';

// Constants
import {
  DEFAULT_CURRENT_PAGE,
  DEFAULT_PAGE_SIZE,
  DEFAULT_TOTAL,
  OPTION_LIMITS,
} from '@/constants';

// Utils
import { getRecordRange } from '@/utils';

// Components
import { Button, Select } from '../common';

interface PaginationProps {
  totalRecords?: number;
  limit?: number;
  currentPage?: number;
  isDisabledPrev?: boolean;
  isDisabledNext?: boolean;
  onPageChange?: (direction: number | 'prev' | 'next') => void;
  onChangeLimit?: (value: number) => void;
}

const Pagination = ({
  currentPage = DEFAULT_CURRENT_PAGE,
  limit = DEFAULT_PAGE_SIZE,
  totalRecords = DEFAULT_TOTAL,
  isDisabledPrev,
  isDisabledNext,
  onPageChange,
  onChangeLimit,
}: PaginationProps) => {
  const handleNextPage = () => onPageChange?.('next');

  const handlePrevPage = () => onPageChange?.('prev');

  const handleChangeLimit = (value: string) => {
    onChangeLimit?.(Number(value));
  };

  return (
    <div className="w-full flex justify-between mt-6">
      <div className="flex items-center">
        <p className="text-md">
          {getRecordRange(currentPage, limit, totalRecords)}
        </p>
      </div>
      <div className="flex items-center gap-10">
        <div className="flex items-center gap-3 text-md">
          <p>Rows per page:</p>
          <Select
            defaultValue={String(limit)}
            option={OPTION_LIMITS}
            isDisable={!totalRecords}
            onChange={handleChangeLimit}
            className="bg-white border border-gray-dark rounded-md w-fit"
          />
        </div>
        <div className="flex items-center gap-2">
          <Button
            variant="ghost"
            title="prev-button"
            className="bg-primary rounded-md w-7 h-7 p-0"
            disabled={isDisabledPrev}
            onClick={handlePrevPage}
          >
            <MoveLeft className="text-white w-4 h-4" />
          </Button>
          <Button
            variant="ghost"
            title="next-button"
            className="bg-primary rounded-md w-7 h-7 p-0"
            disabled={isDisabledNext}
            onClick={handleNextPage}
          >
            <MoveRight className="text-white w-4 h-4" />
          </Button>
        </div>
      </div>
    </div>
  );
};

export default memo(Pagination, isEqual);
