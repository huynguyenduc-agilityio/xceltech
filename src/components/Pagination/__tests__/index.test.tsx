import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import Pagination from '..';

const onPageChangeMock = jest.fn();
const onChangeLimitMock = jest.fn();

describe('Pagination render', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <Pagination
        currentPage={1}
        isDisabledNext={true}
        isDisabledPrev={true}
        onChangeLimit={onChangeLimitMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Handle next page', async () => {
    const { getByRole } = render(
      <Pagination onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByRole('button', { name: /next/i });

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle prev page', async () => {
    const { getByRole } = render(
      <Pagination currentPage={2} onPageChange={onPageChangeMock} />,
    );

    const prevPage = getByRole('button', { name: /prev/i });

    await userEvent.click(prevPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('should render default value correctly', () => {
    const { container } = render(<Pagination />);

    expect(container).toBeInTheDocument();
  });
});
