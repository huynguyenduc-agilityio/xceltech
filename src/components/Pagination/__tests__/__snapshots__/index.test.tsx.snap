// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render should match snapshot 1`] = `
<div>
  <div
    class="w-full flex justify-between mt-6"
  >
    <div
      class="flex items-center"
    >
      <p
        class="text-md"
      >
        1 - 0 of 0 Leaves
      </p>
    </div>
    <div
      class="flex items-center gap-10"
    >
      <div
        class="flex items-center gap-3 text-md"
      >
        <p>
          Rows per page:
        </p>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-label="Select"
          class="font-productsans flex h-full items-center justify-between px-3 py-2 text-lg ring-offset-background data-[placeholder]:text-gray-soft focus:outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 bg-white border border-gray-dark rounded-md w-fit"
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            5
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex items-center gap-2"
      >
        <button
          class="font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed hover:bg-accent hover:text-accent-foreground bg-primary rounded-md w-7 h-7 p-0"
          disabled=""
          title="prev-button"
        >
          <svg
            class="lucide lucide-move-left text-white w-4 h-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 8L2 12L6 16"
            />
            <path
              d="M2 12H22"
            />
          </svg>
        </button>
        <button
          class="font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed hover:bg-accent hover:text-accent-foreground bg-primary rounded-md w-7 h-7 p-0"
          disabled=""
          title="next-button"
        >
          <svg
            class="lucide lucide-move-right text-white w-4 h-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 8L22 12L18 16"
            />
            <path
              d="M2 12H22"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
